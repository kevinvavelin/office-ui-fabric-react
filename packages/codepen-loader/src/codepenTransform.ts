import * as babylon from 'babylon';
import * as prettier from 'prettier';

import { Recast, JSCodeShift, IASTPath, IASTSpecifier, IASTNode } from './interfaces';

// recast and jscodeshift have no @types :(
const recast: Recast = require('recast');
const jscodeshift: {
  withParser: (parser: string) => JSCodeShift;
} = require('jscodeshift');

const j = jscodeshift.withParser('babylon');

const parse = (source: string) =>
  babylon.parse(source, {
    sourceType: 'module',
    plugins: ['jsx', 'typescript', 'classProperties', 'objectRestSpread']
  });

/**
 * This Transform modifies Fabric website code examples into a format that will allow them to be
 * rendered on Codepen, as part of the "Export to Codepen" feature.
 * There are two types of supported example templates:
 *
 *     [imports]
 *     [named variable export with example code inside]
 *
 * and
 *
 *     [imports]
 *     [named class export with example code inside]
 *
 * Currently, examples which are scattered across multiple files are NOT supported.
 *
 * @param file Example file contents to transform
 * @returns The transformed file contents
 */
export function transform(file: string): string {
  const sourceStr = file;
  const source = j(recast.parse(sourceStr, { parser: { parse } }));

  // Make a list of imported identifiers (normal and for example data), and remove all imports
  const identifiers: string[] = [];
  const exampleIdentifiers: string[] = [];
  source.find(j.ImportDeclaration).forEach((p: IASTPath) => {
    const importPath = p.node.source.value;
    // Ignore identifiers from:
    // - the React import (which will be a global)
    // - css/scss
    if (importPath !== 'react' && !/\.s?css$/.test(importPath)) {
      const isExample = importPath.startsWith('@uifabric/example-data');
      p.node.specifiers.forEach((spec: IASTSpecifier) => {
        if (isExample) {
          exampleIdentifiers.push(spec.local.loc.identifierName);
        } else {
          identifiers.push(spec.local.loc.identifierName);
        }
      });
    }

    // Remove the import
    p.prune();
  });

  let exampleName;
  // remove exports and replace with variable or class declarations, whichever the original example used
  source
    .find(
      j.ExportNamedDeclaration,
      (node: IASTNode) => node.declaration.type === 'VariableDeclaration' || node.declaration.type === 'FunctionDeclaration'
    )
    .replaceWith((p: IASTPath) => {
      if (p.node.declaration.type === 'VariableDeclaration') {
        exampleName = p.node.declaration.declarations[0].id.name;
      }
      return p.node.declaration;
    });

  source
    .find(
      j.ExportNamedDeclaration,
      (node: IASTNode) => node.declaration.type === 'ClassDeclaration' || node.declaration.type === 'TSInterfaceDeclaration'
    )
    .replaceWith((p: IASTPath) => {
      if (p.node.declaration.type === 'ClassDeclaration') {
        exampleName = p.node.declaration.id.name;
      }
      return p.node.declaration;
    });

  // add imports and React render footer (with the component wrapped in a <Fabric> for styling)
  const sourceWithFooter = [source.toSource(), `ReactDOM.render(<Fabric><${exampleName}/></Fabric>, document.getElementById("content"));`];

  // Build the list of imports from window.Fabric and window.FabricExampleData
  if (exampleIdentifiers.length) {
    sourceWithFooter.unshift(`const {${exampleIdentifiers.join(',')}} = window.FabricExampleData;`);
  }
  if (!identifiers.includes('Fabric')) {
    identifiers.push('Fabric');
  }
  sourceWithFooter.unshift(`const {${identifiers.join(',')}} = window.Fabric;`);

  return prettier.format(sourceWithFooter.join('\n'), {
    parser: 'typescript',
    printWidth: 100,
    tabWidth: 2,
    singleQuote: true
  });
}
