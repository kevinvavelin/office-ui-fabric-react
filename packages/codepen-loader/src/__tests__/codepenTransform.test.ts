import { transform } from '../codepenTransform';
import * as fs from 'fs';
import * as path from 'path';

describe('codepen transform', () => {
  function transformFile(file: string): string {
    const filename = path.resolve(__dirname, './examples/' + file);
    const fileContents = fs.readFileSync(filename).toString();
    return transform(fileContents);
  }

  it('handles examples with function components', () => {
    const result = transformFile('function.txt');
    expect(result).toMatchSnapshot();
  });

  it('handles examples with class components', () => {
    const result = transformFile('class.txt');
    expect(result).toMatchSnapshot();
  });

  it('handles examples importing exampleData', () => {
    const result = transformFile('exampleData.txt');
    expect(result).toMatchSnapshot();
  });
});
