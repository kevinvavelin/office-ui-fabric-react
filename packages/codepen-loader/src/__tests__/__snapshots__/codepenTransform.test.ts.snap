// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codepen transform handles examples importing exampleData 1`] = `
"const { Fabric } = window.Fabric;
const { createListItems, IExampleItem } = window.FabricExampleData;
const FakeExampleDataExample = () => {
  return <div>hi</div>;
};

ReactDOM.render(
  <Fabric>
    <FakeExampleDataExample />
  </Fabric>,
  document.getElementById('content')
);
"
`;

exports[`codepen transform handles examples with class components 1`] = `
"const { SpinButton, Fabric } = window.Fabric;
class SpinButtonBasicExample extends React.Component<any, any> {
  public render(): JSX.Element {
    return (
      <div style={{ width: '400px' }}>
        <SpinButton
          defaultValue=\\"0\\"
          label={'Basic SpinButton:'}
          min={0}
          max={100}
          step={1}
          iconProps={{ iconName: 'IncreaseIndentLegacy' }}
          incrementButtonAriaLabel={'Increase value by 1'}
          decrementButtonAriaLabel={'Decrease value by 1'}
        />
      </div>
    );
  }
}

ReactDOM.render(
  <Fabric>
    <SpinButtonBasicExample />
  </Fabric>,
  document.getElementById('content')
);
"
`;

exports[`codepen transform handles examples with function components 1`] = `
"const { Label, Fabric } = window.Fabric;
const LabelBasicExample = () => {
  return (
    <div>
      <Label>I'm a Label</Label>
    </div>
  );
};

ReactDOM.render(
  <Fabric>
    <LabelBasicExample />
  </Fabric>,
  document.getElementById('content')
);
"
`;
